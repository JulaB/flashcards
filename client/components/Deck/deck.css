.deck {
  border-radius: 5px;
  border: 1px solid var(--color-border);

  &__title {
    margin-top: 0;
    text-align: center;
    font-weight: normal;
    background: var(--color-card-bg);
    padding: var(--space-sm);
    color: var(--color-text-light);
  }

  &__info {
    padding: var(--space-sm);
  }
}
