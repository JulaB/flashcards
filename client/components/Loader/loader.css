@keyframes dot-bounce {
  0%,
  100% { transform: scale(0); }
  50% { transform: scale(1); }
}

.loader {
  display: flex;
  margin: 0 auto;
  width: 66px;

  &__dot1,
  &__dot2,
  &__dot3 {
    width: 16px;
    height: 16px;
    margin: 3px;
    border-radius: 50%;
    background-color: var(--color-loader);
    animation: dot-bounce 1.4s ease-in-out infinite;
  }

  &__dot2 {
    animation-delay: 0.16s;
  }

  &__dot3 {
    animation-delay: 0.32s;
  }

  &--btn {
    width: auto;

    .loader__dot1,
    .loader__dot2,
    .loader__dot3 {
      width: calc(var(--text-base-size) - 6px);
      height: calc(var(--text-base-size) - 6px);
      background-color: var(--btn-primary-text);
    }
  }
}
