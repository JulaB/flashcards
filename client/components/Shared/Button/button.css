:root {
  --btn-font-size: 1em;
  --btn-radius: 9999px;
}

.btn {
  text-decoration: none;
  font-size: var(--btn-font-size);
  line-height: 1;
  border: 0;
  border-radius: var(--btn-radius);
  color: var(--btn-default-text);
  padding: 10px 15px;
  outline: none;
  background: var(--btn-default-bg);
  cursor: pointer;
  display: inline-flex;
  min-height: var(--text-base-size);

  &--primary {
    background: var(--btn-primary-bg);
    border: 0;
    color: var(--btn-primary-text);
  }

  &--primary:hover,
  &--primary:focus {
    background: var(--btn-primary-hover-bg);
    color: var(--btn-primary-hover-text);
  }

  &--primary:disabled {
    opacity: 0.7;
    cursor: not-allowed;
    pointer-events: none;
  }
}
