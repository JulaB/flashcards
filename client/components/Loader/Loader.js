import React from 'react';
import PropTypes from 'prop-types';
import './loader.css';

const Loader = ({ className }) => (
  <div className={`loader ${className}`}>
    <div className="loader__dot1" />
    <div className="loader__dot2" />
    <div className="loader__dot3" />
  </div>
);

Loader.propTypes = {
  className: PropTypes.string,
};

Loader.defaultProps = {
  className: '',
};

export default Loader;
