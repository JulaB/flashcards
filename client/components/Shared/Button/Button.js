import React from 'react';
import PropTypes from 'prop-types';

import './button.css';

const Button = ({
  className,
  type,
  children,
  ...props
}) => (
  <button type={type} className={`btn ${className}`} {...props}>
    {children}
  </button>
);

Button.propTypes = {
  className: PropTypes.string,
  type: PropTypes.string,
  children: PropTypes.node,
};

Button.defaultProps = {
  className: '',
  type: 'button',
  children: null,
};

export default Button;
