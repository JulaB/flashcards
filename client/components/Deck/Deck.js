import React from 'react';
import PropTypes from 'prop-types';
import Pluralize from 'components/Shared/Pluralize';
import Button from 'components/Shared/Button/Button';

import './deck.css';

const Deck = ({ name, id, ...props }) => (
  <div className="deck">
    <h4 className="deck__title">{name}</h4>
    <div className="deck__info">
      <Pluralize singular='card' plural='cards' count={props['cards-count']} />
      <Button className="btn--primary deck__button">Quick View</Button>
    </div>
  </div>
);

Deck.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  'cards-count': PropTypes.number,
};

Deck.defaultProps = {
  id: '0',
  name: '',
  'cards-count': 0,
};
export default Deck;
