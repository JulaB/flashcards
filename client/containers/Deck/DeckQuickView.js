import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Loader from 'components/Loader/Loader';
import Button from 'components/Shared/Button/Button';
import { deckFetchData as fetchData } from 'actions/deckActions';
import { showModal } from 'actions/modalActions';
import { DECK_MODAL } from 'constants/modalConstants';

export class DeckQuickViewContainer extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    const { isFetching, deckId } = this.props;
    if (isFetching) {
      return;
    }
    this.props.openModalFor(deckId);
  }

  render() {
    const { isFetching } = this.props;
    return (
      <Button
        className="btn--primary deck__button"
        onClick={this.handleClick}
        disabled={isFetching}
      >
        { isFetching
          ? <Loader className="loader--btn" />
          : 'Quick View'
        }
      </Button>
    );
  }
}

DeckQuickViewContainer.propTypes = {
  deckId: PropTypes.string.isRequired,
  name: PropTypes.string,
  isFetching: PropTypes.bool,
  openModalFor: PropTypes.func,
};

DeckQuickViewContainer.defaultProps = {
  isFetching: false,
  name: '',
  openModalFor: () => {},
};

const mapStateToProps = ({ deckReducer: state }, ownProps) => ({
  isFetching: state.deckIsFetching[ownProps.deckId] || false,
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  openModalFor: deckId => (
    dispatch(fetchData(deckId))
      .then(() => dispatch(
        showModal({
          modalType: DECK_MODAL,
          modalProps: {
            deckId,
            open: true,
            title: ownProps.name,
          },
        }),
      ))
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(DeckQuickViewContainer);
