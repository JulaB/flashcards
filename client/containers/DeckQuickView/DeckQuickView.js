import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { deckFetchData as fetchData } from 'actions/deckActions';
import Loader from 'components/Loader/Loader';
import Button from 'components/Shared/Button/Button';
// import Slider from 'components/Shared/Slider/Slider';
// import Card from 'components/Card/Card';

export class DeckQuickViewContainer extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    const { isFetching, deckId } = this.props;
    if (isFetching) {
      return;
    }
    this.props.fetchData(deckId);
  }

  render() {
    const { isFetching } = this.props;
    return (
      <Button className="btn--primary deck__button" onClick={this.handleClick}>
        { isFetching
          ? <Loader className="loader--btn" />
          : 'Quick View'
        }
      </Button>
    );
    // return (
    //   <Slider>
    //     {cards.map(card => (
    //       <Card {...card.attributes} key={card.id} />
    //     ))}
    //   </Slider>
    // );
  }
}

DeckQuickViewContainer.propTypes = {
  deckId: PropTypes.string.isRequired,
  isFetching: PropTypes.bool,
  cards: PropTypes.arrayOf(PropTypes.object),
  fetchData: PropTypes.func,
};

DeckQuickViewContainer.defaultProps = {
  isFetching: false,
  cards: [],
  fetchData: () => {},
};

const mapStateToProps = ({ deckReducer: state }, ownProps) => ({
  isFetching: state.deckIsFetching[ownProps.deckId] || false,
  cards: state.cards[ownProps.deckId] || [],
});

export default connect(mapStateToProps, { fetchData })(DeckQuickViewContainer);
