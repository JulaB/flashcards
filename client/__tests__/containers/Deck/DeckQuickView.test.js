import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { DeckQuickViewContainer as DeckQuickView } from 'containers/Deck/DeckQuickView';
import Loader from 'components/Loader/Loader';
import Button from 'components/Shared/Button/Button';

describe('<DeckQuickView />', () => {
  const deckId = '1';
  it('renders the component with default props', () => {
    const wrapper = shallow(<DeckQuickView deckId={deckId} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders loader when isFetching prop is set', () => {
    const wrapper = shallow(<DeckQuickView deckId={deckId} isFetching />);
    expect(wrapper.find(Loader)).toHaveLength(1);
    expect(wrapper.find(Button).text()).not.toContain('Quick View');
  });

  it('cannot double click button while fetching', () => {
    const openModalFor = jest.fn();
    const wrapper = shallow(
      <DeckQuickView
        deckId={deckId}
        isFetching
        openModalFor={openModalFor}
      />,
    );
    wrapper.find(Button).simulate('click');
    expect(openModalFor).not.toHaveBeenCalled();
  });

  it('responds to click', () => {
    const openModalFor = jest.fn();
    const wrapper = shallow(
      <DeckQuickView
        deckId={deckId}
        openModalFor={openModalFor}
      />,
    );
    wrapper.find(Button).simulate('click');
    expect(openModalFor).toHaveBeenCalledWith(deckId);
  });
});
