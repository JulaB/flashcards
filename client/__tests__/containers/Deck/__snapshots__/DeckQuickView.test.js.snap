// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeckQuickView /> renders the component with default props 1`] = `
<Button
  className="btn--primary deck__button"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Quick View
</Button>
`;
